runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,482 @@
+        local index = math.max(unpack((#tabstop_indexes == 0 and { 0 }) or tabstop_indexes)) + 1
+        add_tabstop(index, value)
+      end
+      append_to_snippet(snippet_text, value)
+    elseif type == G.NodeType.Text then
+      --- @cast data vim.snippet.TextData
+      append_to_snippet(snippet_text, data.text)
+    end
+  end
+
+  -- $0, which defaults to the end of the snippet, defines the final cursor position.
+  -- Make sure the snippet has exactly one of these.
+  if vim.tbl_contains(vim.tbl_keys(tabstop_ranges), 0) then
+    assert(#tabstop_ranges[0] == 1, 'Snippet has multiple $0 tabstops')
+  else
+    add_tabstop(0)
+  end
+
+  -- Insert the snippet text.
+  local cursor_row, cursor_col = cursor_pos()
+  vim.api.nvim_buf_set_text(
+    M.session.bufnr,
+    cursor_row,
+    cursor_col,
+    cursor_row,
+    cursor_col,
+    remove_newlines(snippet_text)
+  )
+
+  -- Create the tabstop extmarks.
+  for index, ranges in pairs(tabstop_ranges) do
+    for _, range in ipairs(ranges) do
+      M.session:add_tabstop(index, range)
+    end
+  end
+
+  -- Jump to the first tabstop.
+  M.jump(1)
+end
+
+--- @alias vim.snippet.Direction
+--- | -1 # Previous
+--- | 1 # Next
+
+--- Returns `true` if there is an active snippet which can be jumped in the given direction.
+---
+--- @param direction vim.snippet.Direction
+--- @return boolean
+function M.jumpable(direction)
+  if not M.active() or M.session:get_num_tabstops() == 0 then
+    return false
+  end
+
+  if direction == -1 then
+    return M.session.current_tabstop.index > 1
+  else
+    return M.session.current_tabstop.index < M.session:get_num_tabstops()
+  end
+end
+
+--- Jumps (if possible) within the active snippet in the given direction.
+---
+--- @param direction vim.snippet.Direction
+function M.jump(direction)
+  -- TODO: Not sure if we want to have this check or insist on the user to check `jumpable` first.
+  if not M.jumpable(direction) then
+    return
+  end
+
+  -- Find the tabstop with the lowest range.
+  local tabstops =
+    M.session.tabstops[(M.session.current_tabstop.index + direction) % M.session:get_num_tabstops()]
+  local dest = tabstops[1]
+  for _, tabstop in ipairs(tabstops) do
+    local dest_range, range = dest:get_range(), tabstop:get_range()
+    if (range[1] < dest_range[1]) or (range[1] == dest_range[1] and range[2] < dest_range[2]) then
+      dest = tabstop
+    end
+  end
+
+  -- Clear the autocommands so that we can move the cursor freely while selecting the tabstop.
+  M._clear_autocmds()
+
+  M.session.current_tabstop = dest
+  M._select_stop(dest)
+
+  -- Restore the autocommands.
+  M._setup_autocmds()
+end
+
+--- Returns `true` if there's an active snippet in the current buffer.
+---
+--- @return boolean
+function M.active()
+  return M.session ~= nil and M.session.bufnr == vim.api.nvim_get_current_buf()
+end
+
+-- TODO: Use another name? discard? dispose? unlink? assassinate?
+--- Cancels the current snippet.
+function M.cancel()
```

Comment:
I think cancel is a good match, and it has the most uses across the codebase:

```
~/dev/neovim/neovim master
↪  rg cancel | wc -l
416

~/dev/neovim/neovim master
↪  rg discard | wc -l
311

~/dev/neovim/neovim master
↪  rg unlink | wc -l
39

~/dev/neovim/neovim master
↪  rg dispose | wc -l
3
```

Reply:
Oh, I had missed this:

> since sometimes it does cancel the snippet session but it's also invoked when the snippet navigation is complete. Do we want to give it another name?

In that case, maybe `reset` could work?

It's also used quite often:

```
↪  rg reset | wc -l
1160
```

Reply:
I still prefer @clason's suggestion of `exit`. I feel like `reset` doesn't imply that the snippet will be discarded entirely, instead of just reinitialized or something.